### PR TITLE
Added count_documents() implementation to builtin_timeseries

### PR DESCRIPTION
**Pull request by MukuFlash03**
Implemented code for issue 933 (you can replace this with an link https://github.com/e-mission/e-mission-docs/issues/933, @MukuFlash03 ) in e-mission-docs for adding functionality to count number of documents. I've determined that 'key' parameter can be passed to retrieve appropriate timeseries db collection. A query is generated with optional extra_query keys list which returns filtered data set.

Pending: Overall testing with both regular and edge test cases.

### CODE REVIEW

**Comment by nataliejschultz** [comment]
[Link to the issue ](https://github.com/e-mission/e-mission-docs/issues/933)

File: emission/tests/storageTests/TestTimeSeries.py [hunk 1]
```diff
@@ -81,6 +81,22 @@ def testExtraQueries(self):
...
+        Test 1 : Specific key with empty extra_queries
+        key = 'background/location', extra_query_list = []
+        Results in empty query = {}, which matches all documents for a user for that key.
+        Hence should return total count of all documents matching that key.
+        Testing this with sample dataset: "shankari_2015-aug-27"
+        '''
+        ts = esta.TimeSeries.get_time_series(self.testUUID)
+        total_count = ts.count_data("background/location",[])
+        print(total_count)
+        self.assertEqual(total_count, 555)
```
Comment (shankari) [inline emission/tests/storageTests/TestTimeSeries.py:95]: How do you know that 555 is correct?

**Review by shankari — CHANGES_REQUESTED** [review]
Looks good! Only one minor change...

File: emission/storage/timeseries/builtin_timeseries.py [hunk 2]
```diff
@@ -440,3 +440,15 @@ def update_data(user_id, key, obj_id, data):
         logging.debug("updating entry %s into timeseries" % new_entry)
         edb.save(ts.get_timeseries_db(key), new_entry)
 
+    def count_data(self, key, extra_query_list):
```
Comment (shankari) [inline emission/storage/timeseries/builtin_timeseries.py:443]: nit: is `count_data` really the best term? we have `find_entries`, so we might want to make this more consistent instead of coming up with Yet Another Name.

File: emission/tests/storageTests/TestTimeSeries.py (hunk 1 above)
Comment (MukuFlash03) [reply emission/tests/storageTests/TestTimeSeries.py:95]: With reference to my [notes on this issue](https://github.com/e-mission/e-mission-docs/issues/933#issuecomment-1698041613), I realized that the sample dataset `shankari_2015-aug-27` is the dataset on which the count functionality is tested in the test scripts.

So, I did a manual count of the number of occurrences of the key values: `"background/location"` and `"background/filtered_location"` using the grep commands:
- `grep -c "background/location" shankari_2015-aug-27.json`
- `grep -c "background/filtered_location" shankari_2015-aug-27.json`
- This returned the counts as 555 and 327, respectively.

I did assume the count is required for input data which is the one that gets stored in timeseries_db() or analysis_timeseries_db().

**If this approach is indeed correct, I can add the test input for "background/filtered_location" as well.**

Attached is the screenshot for the grep results.

Screen Shot 2023-08-31 at 11 19 52 AM

Comment (shankari) [reply emission/tests/storageTests/TestTimeSeries.py:95]: This is a fine approach, but:
1. screenshots for code are bad because they are not searchable
2. put the justification *in the code*. How is somebody looking at this code 5 years later going to know how you validated the answer?

Comment (shankari) [reply emission/tests/storageTests/TestTimeSeries.py:95]: Also, 

> I realized that the sample dataset shankari_2015-aug-27 is the dataset on which the count functionality is tested in the test scripts.

is incorrect. That is a suggestion on loading data for **manual** end-to-end testing. You cannot use manual instructions for automated testing - the test has to be self-contained and load all the data that it needs. This set of tests does load its own data already but not from that file.

File: emission/tests/storageTests/TestTimeSeries.py [hunk 3]
```diff
@@ -81,6 +81,30 @@ def testExtraQueries(self):
...
+            - Validated using grep count of occurrences for keys: 1) "background/location"     2) "background/filtered_location"
+                - $ grep -c <key> <dataset>.json
+        '''
+        ts1_aug_21 = esta.TimeSeries.get_time_series(self.testUUID1)
+        ts2_aug_27 = esta.TimeSeries.get_time_series(self.testUUID)
+
+        count_ts1 = [ts1_aug_21.find_entries_count(key="background/location"), ts1_aug_21.find_entries_count(key="background/filtered_location")]
+        print("\nEntry counts for location, filtered_location on {} = {}".format("Aug_21", count_ts1))
+        self.assertEqual(count_ts1, [738, 508])
+
```
Comment (shankari) [inline emission/tests/storageTests/TestTimeSeries.py:100]: These are very basic tests - I don't see any corner-case testing, for example.
- What happens if the key list is blank?
- What happens if the key list is `blablabla`?
- What happens if there is no data for that user?

I believe you had listed out a few similar cases during our discussion, I don't see any actual implementation of those here.

**Review by shankari — CHANGES_REQUESTED** [review]

File: emission/tests/storageTests/TestTimeSeries.py (hunk 3 above)
Comment (MukuFlash03) [reply emission/tests/storageTests/TestTimeSeries.py:100]: I had not implemented for blank keys as it was to be decided whether key is to be made mandatory or not and assumed that a key would always be passed for getting count. 
But yes, I will go ahead and make code changes to accommodate for blank keys by returning total counts from each of the two timeseries DBs as a tuple (count_timeseries_db, count_analysis_db).

Comment (MukuFlash03) [reply emission/tests/storageTests/TestTimeSeries.py:100]: I had a doubt, regarding `"key list"`. For now, I have only been taking a single key as opposed to a list.
1) Is the count to be done for multiple keys passed as a key list?
2) If yes, ´get_query'  returns a combined query for multiple keys and gives total count of matching entries not segregated by keys. For instance, keylist = [A,B], then total count will be (A+B).
Do we need a separate count for each key in that case, that is (A,B) or combined count = (A+B) is needed?

**Comment by shankari** [comment]
> I had not implemented for blank keys as it was to be decided whether key is to be made mandatory or not and assumed that a key would always be passed for getting count.

Let me clarify.

It is fine to assume that a key needs to be passed in for getting count.
Does everybody read documentation and follow it perfectly? No.

But we should also define the behavior if it is not **and test that behavior**
That is pretty much the definition of unit testing

File: emission/tests/storageTests/TestTimeSeries.py (hunk 3 above)
Comment (shankari) [reply emission/tests/storageTests/TestTimeSeries.py:100]: Please see https://github.com/e-mission/e-mission-server/pull/935#issuecomment-1707079925

**Comment by MukuFlash03** [comment]
With reference to the TestTimeSeries.py file's testcase for Aggregate subclass under the testFindEntriesCount module,
I am getting a Failed assertion for this test.

**Problem**: 
Test passes on running only the specific test module or even the specific test file. 
However, test fails when the entire suit of tests is run using runAllTests.sh both locally and through the GitHub CLI.

**Possible issue**:
I am fetching the distinct users under the aggregate class object, and for each user - fetching the timeseries and counting documents.

In case of running only current test module, only 1 distinct user is found and output matches expected counts. I have verified expected outputs using mongodb terminal to fetch the count values. Code comments contains details on my validation process.

On the entire test suite, this user may not be present or the count data may not be matching for the user.

-----------
**Failed Assertion:**

 ```
 File "/Users/mmahadik/Documents/GitHub/e-mission-server/emission/tests/storageTests/TestTimeSeries.py", line 156, in testFindEntriesCount
    self.assertEqual(count_ts7, ([1476, 1016], [5]))
AssertionError: Tuples differ: ([0, 0], [0]) != ([1476, 1016], [5])

First differing element 0:
[0, 0]
[1476, 1016]

- ([0, 0], [0])
+ ([1476, 1016], [5])

----------------------------------------------------------------------
Ran 345 tests in 1327.825s
FAILED (failures=1)
```

File: emission/storage/timeseries/builtin_timeseries.py [hunk 4]
```diff
@@ -440,3 +440,63 @@ def update_data(user_id, key, obj_id, data):
...
+
+        if key_list == [] or key_list is None:
+            key_list = None
+        
+        # Segregate orig_tsdb and analysis_tsdb keys
+        (orig_tsdb_keys, analysis_tsdb_keys) = self._split_key_list(key_list)
+
+        orig_tsdb_counts = self._get_entries_counts_for_timeseries(orig_tsdb, orig_tsdb_keys, time_query, geo_query, extra_query_list)
+        analysis_tsdb_counts = self._get_entries_counts_for_timeseries(analysis_tsdb, analysis_tsdb_keys, time_query, geo_query, extra_query_list)
+
```
Comment (shankari) [inline emission/storage/timeseries/builtin_timeseries.py:477]: I am not opposed to this, but I thought that the plan was to only support one key for now so that we could make progress on other server tasks. Can you clarify why that changed?

File: emission/tests/storageTests/TestTimeSeries.py [hunk 5]
```diff
@@ -81,6 +81,76 @@ def testExtraQueries(self):
...
 
+    def testFindEntriesCount(self):
+        '''
+        Test: Specific keys with other parameters not passed values.
+        Input: For each dataset: ["background/location", "background/filtered_location", "analysis/confirmed_trip"]
+            - Testing this with sample dataset: "shankari_2015-aug-21", "shankari_2015-aug-27"
+        Output: Aug_21: ([738, 508], [0]), Aug_27: ([555, 327], [0])
+            - Actual output just returns a single number for count of entries.
+            - Validated using grep count of occurrences for keys: 1) "background/location"     2) "background/filtered_location"
+                - $ grep -c <key> <dataset>.json
```
Comment (shankari) [inline emission/tests/storageTests/TestTimeSeries.py:92]: can you please add the `grep` outputs here?

File: emission/tests/storageTests/TestTimeSeries.py [hunk 6]
```diff
@@ -81,6 +81,76 @@ def testExtraQueries(self):
...
+
+        '''
+
+        ts1_aug_21 = esta.TimeSeries.get_time_series(self.testUUID1)
+        ts2_aug_27 = esta.TimeSeries.get_time_series(self.testUUID)
+
+        # Test case: Combination of original and analysis timeseries DB keys for Aug-21 dataset
+        key_list1=["background/location", "background/filtered_location", "analysis/confirmed_trip"]
+        count_ts1 = ts1_aug_21.find_entries_count(key_list=key_list1)
+        self.assertEqual(count_ts1, ([738, 508], [0]))
```
Comment (shankari) [inline emission/tests/storageTests/TestTimeSeries.py:110]: why is this not `[0,0]` to match the `2x1` array from the timeseries for the two keys?

File: emission/tests/storageTests/TestTimeSeries.py [hunk 7]
```diff
@@ -81,6 +81,76 @@ def testExtraQueries(self):
...
+
+        # Test case: Combination of original and analysis timeseries DB keys for Aug-27 dataset
+        key_list1=["background/location", "background/filtered_location", "analysis/confirmed_trip"]
+        count_ts2 = ts2_aug_27.find_entries_count(key_list=key_list1)
+        self.assertEqual(count_ts2, ([555, 327], [0]))
```
Comment (shankari) [inline emission/tests/storageTests/TestTimeSeries.py:115]: Ditto

File: emission/tests/storageTests/TestTimeSeries.py [hunk 8]
```diff
@@ -81,6 +81,76 @@ def testExtraQueries(self):
...
+
+        # Test case: Only original timeseries DB keys for Aug-27 dataset
+        key_list2=["background/location", "background/filtered_location"]
+        count_ts3 = ts2_aug_27.find_entries_count(key_list=key_list2)
+        self.assertEqual(count_ts3, ([555, 327], []))
```
Comment (shankari) [inline emission/tests/storageTests/TestTimeSeries.py:120]: And here it's an empty array. Why the inconsistency?

File: miniconda.sh
Comment (shankari) [inline miniconda.sh:1]: why is this file checked in?

**Review by shankari — CHANGES_REQUESTED** [review]
Better, but still needs some work.

File: emission/storage/timeseries/builtin_timeseries.py (hunk 4 above)
Comment (MukuFlash03) [reply emission/storage/timeseries/builtin_timeseries.py:477]: The blank key input scenario prompted me to go for supporting multiple keys.
In this case, I thought of returning total counts of all documents for the specific user not filtered by keys.

Since, this is a functionality of abstract_timeseries class and its subclasses, any function call for counting documents would happen through this class objects and should not directly involve any timeseries database.

Now, since we don't know what timeseries database is to be used for counting, my thought was to return count of documents pertaining to both timeseries databases for the user (original and analysis).

I do realize now, that this could still have been achieved with a single key instead of key_list but a series of code changes happened in my current thought process to accommodate code changes and new test cases. 

The extra time spent on the modifying the initial requirement could have been utilized for the other pending tasks.

File: emission/tests/storageTests/TestTimeSeries.py (hunk 6 above)
Comment (MukuFlash03) [reply emission/tests/storageTests/TestTimeSeries.py:110]: The array size depends on the number of keys pertaining to either of the two timeseries databases.
`background keys` are two in number, hence two array elements.
Similarly `analysis keys` are one in total, hence just a single element.
If two analysis keys were present, then the returned array list would have two entries for each analysis key.

File: emission/tests/storageTests/TestTimeSeries.py (hunk 7 above)
Comment (MukuFlash03) [reply emission/tests/storageTests/TestTimeSeries.py:115]: Answered [here](https://github.com/e-mission/e-mission-server/pull/935#discussion_r1317914465)

File: emission/tests/storageTests/TestTimeSeries.py (hunk 8 above)
Comment (MukuFlash03) [reply emission/tests/storageTests/TestTimeSeries.py:120]: Empty array is returned in case there were no keys pertaining to the respective timeseries database.
In this specific test case, only background keys are there, hence their count is returned.
No, analysis keys are present in the input keys, hence empty array returned.

This is to differentiate from the [0] case where a key might be present in the input but no matching documents found.
Whereas in this case of [], no key was present in the input itself.

File: miniconda.sh
Comment (MukuFlash03) [reply miniconda.sh:1]: Apologies, will remove. I had reset my environment by running `setup/setup.sh`; forgot to remove this file.

File: emission/tests/storageTests/TestTimeSeries.py (hunk 8 above)
Comment (shankari) [reply emission/tests/storageTests/TestTimeSeries.py:135]: That is an interesting design choice, but I am not sure I agree with it. This seems to:
1. **very tied to this specific implementation:** e.g. the difference between timeseries and analysis_timeseries. Whenever you are designing an abstract interface, think about what it would look like if the underlying implementation were completely different. What if we stored everything in giant csv files instead? What if we used a real timeseries database instead? What would the meaning of `timeseries` and `analysis_timeseries` be in that case?
2. **unnecesarily complex to use:** again, think of the user of the interface, who is using the timeseries abstraction that we have outlined in chapter 5 of my thesis. they want to know how many entries there are in each stream. Do they want to know where the entries came from? Do they want to know the details of our implementation? Do they want to know the difference between the key being present but there being zero entries and the key being absent? No. We shouldn't make them do that. Note that with this implementation, we don't even support just adding the two sets of counts together because they may not have the same number of entries.

File: emission/tests/storageTests/TestTimeSeries.py (hunk 7 above)
Comment (shankari) [reply emission/tests/storageTests/TestTimeSeries.py:130]: https://github.com/e-mission/e-mission-server/pull/935#discussion_r1319156496

File: emission/tests/storageTests/TestTimeSeries.py (hunk 6 above)
Comment (shankari) [reply emission/tests/storageTests/TestTimeSeries.py:125]: https://github.com/e-mission/e-mission-server/pull/935#discussion_r1319156496

File: emission/tests/storageTests/TestTimeSeries.py (hunk 8 above)
Comment (MukuFlash03) [reply emission/tests/storageTests/TestTimeSeries.py:135]: Thank for your feedback on the design.

Yes, I do agree that it is too tightly based on the current implementation, especially with regards to two types of timeseries that we have currently.

In terms of reducing the complexity, it again comes down to supporting just a single key as per the initial requirement. This would resolve much of the varying output scenarios that the current implementation has such as the [] vs [0] case.

Motivation for this design:
In the original issue sample code usage [here](https://github.com/AlirezaRa94/op-admin-dashboard/blob/11f5af3e8eae8a54abfb21bc225ab88f6f86b8c7/utils/db_utils.py#L106C1-L122C46), I saw that first the data was fetched and then count was called. However, I designed this keeping in mind that the user would not have to explicitly state which timeseries the user wants a count for. 
I noticed that the key could be used to identify which timeseries is required and hence went ahead with taking in a key.

Pending issue:
Now, in this design scenario, there was the case of blank keys passed. 
**How would we decide what timeseries count was to be returned to the user for this case?** 
Hence, I went ahead with returning the total count of the two available timeseries.
This resulted in my output now changing from a single value to two separate values for the two timeseries.

Possible resolution:
- Stick to a single key and return count.
- Need to decide how blank keys are handled AND,
- Which timeseries data to be considered if user wants count of all documents without specifying key?


Also, I am a bit unclear what you mean by this - _"Note that with this implementation, we don't even support just adding the two sets of counts together because they may not have the same number of entries."_

**Comment by shankari** [comment]
The design should be driven by what is easy for users to use and generalizable. Not by what is easy to implement.

> Motivation for this design:

Motivation for which design? single key or multi-key?

>  How would we decide what timeseries count was to be returned to the user for this case?

I don't see why this is a problem. What does `find_entries` do if the user doesn't specify a key? Do the same things for counts.

> Also, I am a bit unclear what you mean by this - "Note that with this implementation, we don't even support just adding the two sets of counts together because they may not have the same number of entries."

If you always returned the same number of entries (e.g. if `['background/location`, 'background/filtered_location']` returned `([x, y], [0, 0])` then we could do `np.array(retVal[0]) + np.array(retVal[1])` to get the total counts. But we cannot do that with your current design.

Concretely, after this change, what will the `op-admin-dashboard` code that is changed to use this look like?

File: emission/tests/storageTests/TestTimeSeries.py (hunk 8 above)
Comment (shankari) [reply emission/tests/storageTests/TestTimeSeries.py:135]: The design should be driven by what is easy for users to use and generalizable. Not by what is easy to implement.

> Motivation for this design:

Motivation for which design? single key or multi-key?

>  How would we decide what timeseries count was to be returned to the user for this case?

I don't see why this is a problem. What does `find_entries` do if the user doesn't specify a key? Do the same things for counts.

> Also, I am a bit unclear what you mean by this - "Note that with this implementation, we don't even support just adding the two sets of counts together because they may not have the same number of entries."

If you always returned the same number of entries (e.g. if `['background/location`, 'background/filtered_location']` returned `([x, y], [0, 0])` then we could do `np.array(retVal[0]) + np.array(retVal[1])` to get the total counts. But we cannot do that with your current design.

Concretely, after this change, what will the `op-admin-dashboard` code that is changed to use this look like?

Comment (shankari) [reply emission/tests/storageTests/TestTimeSeries.py:135]: This is why it is good, when you come to a decision point like this, to document the alternatives, think through the pros and cons, and document your decision instead of just refactoring.

Comment (MukuFlash03) [reply emission/tests/storageTests/TestTimeSeries.py:135]: > Motivation for which design? single key or multi-key?

Single key initially. 
Approach to fetch timeseries data using keys is valid even in case of multi-key.
I moved to multi-key since I followed how find_entries was handling multiple keys. 

--------

> > How would we decide what timeseries count was to be returned to the user for this case?
> 
> I don't see why this is a problem. What does `find_entries` do if the user doesn't specify a key? Do the same things for counts.

Right, so I have done what `find_entries` is doing - in case of blank keys, fetch the total count of each specific dataset, which itself is done via a helper function `_get_entries_for_timeseries`.
Here, `_get_entries_for_timeseries` (used in `find_entries`) , adds up all counts for keys belonging to a dataset to give a single count.
Ex. ['background/location, 'background/filtered_location'] output would be [x+y]

The difference in my implementation of `_get_entries_counts_for_timeseries` is that I am returning individual counts for each key separately an array element part of combined list.
Ex. ['background/location, 'background/filtered_location'] = [x,y]

This is done for each of the two timeseries and their respective query results are returned in `find_entries`.
Similarly, I am returning the respective counts in `find_entries_count`.

-------

> > Also, I am a bit unclear what you mean by this - "Note that with this implementation, we don't even support just adding the two sets of counts together because they may not have the same number of entries."
> 
> If you always returned the same number of entries (e.g. if `['background/location`, 'background/filtered_location']`returned`([x, y], [0, 0])`then we could do`np.array(retVal[0]) + np.array(retVal[1])` to get the total counts. But we cannot do that with your current design.

Alright, thank you for the clarification. I do get your point. 
This would be a small change to match up the number of entries in the output.

However, I want to point out that the two lists [x,y] and [0,0] would refer to different timeseries datasets with varying keys.
For instance, if 2nd list also had non-zero values, say [a,b], corresponding to some analysis_timseries keys, each element inside the 1st list refers to a separate key which will not match with the key from 2nd list.
Would this be appropriate: `x (= background/location ) + a = (analysis/some_key)  = x + a`


-----

> Concretely, after this change, what will the `op-admin-dashboard` code that is changed to use this look like?

With reference to sample code in original issue [here:](https://github.com/AlirezaRa94/op-admin-dashboard/blob/11f5af3e8eae8a54abfb21bc225ab88f6f86b8c7/utils/db_utils.py#L106C1-L122C46)
`total_trips = edb.get_analysis_timeseries_db().count_documents(arguments include userid and keys... )`,

Now, the count can be fetched by:
`ts_user = esta.get_time_series(UUID(user_id))`
`key_list = ['analysis/confirmed_trip']`
`total_trips = ts_user.find_entries_count(key_list=key_list)[1][0]`

Comment (MukuFlash03) [reply emission/tests/storageTests/TestTimeSeries.py:135]: -----

> Concretely, after this change, what will the `op-admin-dashboard` code that is changed to use this look like?

With reference to sample code in original issue [here:](https://github.com/AlirezaRa94/op-admin-dashboard/blob/11f5af3e8eae8a54abfb21bc225ab88f6f86b8c7/utils/db_utils.py#L106C1-L122C46)
`total_trips = edb.get_analysis_timeseries_db().count_documents(arguments include userid and keys... )`,

Now, the count can be fetched by:
`ts_user = esta.get_time_series(UUID(user_id))`
`key_list = ['analysis/confirmed_trip']`
`total_trips = ts_user.find_entries_count(key_list=key_list)[1][0]`

In case of multiple keys, it would not be easy to access by indexing. So, a better way to return counts would be a dictionary with key-value pair as key_name : count.

Comment (MukuFlash03) [reply emission/tests/storageTests/TestTimeSeries.py:135]: -------

> > Also, I am a bit unclear what you mean by this - "Note that with this implementation, we don't even support just adding the two sets of counts together because they may not have the same number of entries."
> 
> If you always returned the same number of entries (e.g. if `['background/location`, 'background/filtered_location']`returned`([x, y], [0, 0])`then we could do`np.array(retVal[0]) + np.array(retVal[1])` to get the total counts. But we cannot do that with your current design.

Alright, thank you for the clarification. I do get your point. 
This would be a small change to match up the number of entries in the output.

However, I want to point out that the two lists [x,y] and [0,0] would refer to different timeseries datasets with varying keys.
For instance, if 2nd list also had non-zero values, say [a,b], corresponding to some analysis_timseries keys, each element inside the 1st list refers to a separate key which will not match with the key from 2nd list.
Would this be appropriate: `x (= background/location ) + a = (analysis/some_key)  = x + a`

Comment (shankari) [reply emission/tests/storageTests/TestTimeSeries.py:135]: > Right, so I have done what find_entries is doing - in case of blank keys, fetch the total count of each specific dataset, which itself is done via a helper function _get_entries_for_timeseries.
> Here, _get_entries_for_timeseries (used in find_entries) , adds up all counts for keys belonging to a dataset to give a single count.

This is the **implementation** of `find_entries`. The **interface** of `find_entries` is that if you give it blank keys, you get a combined list of all entries in the time range. The count is not even exposed outside the implementation. The **interface** of `find_entry_counts` should be the same as the **interface** of `find_entries`, which is that it should return the number of matching entries for that time range.

> This is done for each of the two timeseries and their respective query results are returned in find_entries.

The respective query results are not returned in `find_entries`. A combined query result is returned.
See `itertools.chain(orig_ts_db_result, analysis_ts_db_result)`

> However, I want to point out that the two lists [x,y] and [0,0] would refer to different timeseries datasets with varying keys.

I fail to see who would be interested in the implementation detail that we have two timeseries collections internally, or why they would need to know which key is stored while accessing the result.

```
ts_user = esta.get_time_series(UUID(user_id))
key_list = ['analysis/confirmed_trip']
total_trips = ts_user.find_entries_count(key_list=key_list)[1][0]
```

and if it were location, we would need:

```
ts_user = esta.get_time_series(UUID(user_id))
key_list = ['background/location']
total_trips = ts_user.find_entries_count(key_list=key_list)[0][0]
```

can you see how that is leaking the implementation outside the interface?

File: emission/tests/storageTests/TestTimeSeries.py [hunk 9]
```diff
@@ -81,6 +81,91 @@ def testExtraQueries(self):
...
+        # Test case: Invalid or unmatched key in metadata field 
+        key_list5=["randomxyz_123test"]
+        with self.assertRaises(KeyError) as ke:
+            count_ts6 = ts1_aug_21.find_entries_count(key_list=key_list5)
+        self.assertEqual(str(ke.exception), "'randomxyz_123test'")
+
+        # Test case: Aggregate timeseries DB User data passed as input
+        ts_agg = esta.TimeSeries.get_aggregate_time_series()
+        count_ts7 = ts_agg.find_entries_count(key_list=key_list1)
+        self.assertEqual(count_ts7, ([1293, 835], [0]))
```
Comment (shankari) [inline emission/tests/storageTests/TestTimeSeries.py:156]:
```suggestion
        self.assertEqual(count_ts7, 2125)
```

**Review by shankari — CHANGES_REQUESTED** [review]
Note that if a list of keys is passed in (e.g. `['background/location', 'background/filtered_location']`, `find_entries` does not currently return lists of entries separated by key. There is no equivalent of `# For each key in orig_tsdb keys, create a query`. It returns a combined list which intersperses both sets of entries. In general, I would expect `find_entries_count(...)` to given the same results as `len(find_entries(...))`

**Comment by MukuFlash03** [comment]
I have addresses the latest review comments.
Overall changes made include using current `_get_entries_for_timeseries` implementation to return the count.
Additionally, changed the returned values to include combined sum for all matching entries.


While the test for the specific test module and the test file runs successfully, the entire test suite fails when run on GitHub's CI as well as my local system.
The failure occurs due to an Assertion failure in the counts for Aggregate subclass test case.
The value is off by 1 and the additional count is for entries from analysis timeseries count.

The actual failed assertion is this:

```
FAIL: testFindEntriesCount (storageTests.TestTimeSeries.TestTimeSeries)
Test: Specific keys with other parameters not passed values.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/e-mission-server/e-mission-server/emission/tests/storageTests/TestTimeSeries.py", line 166, in testFindEntriesCount
    self.assertEqual(count_ts8, 3607)
AssertionError: 3608 != 3607
```

I separated the timeseries counts in the outputs to see which one was not matching, which showed analysis count is getting some entries.

```
FAIL: testFindEntriesCount (storageTests.TestTimeSeries.TestTimeSeries)
Test: Specific keys with other parameters not passed values.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/mmahadik/Documents/GitHub/e-mission-server/emission/tests/storageTests/TestTimeSeries.py", line 174, in testFindEntriesCount
    self.assertEqual(count_ts8, 3607)
AssertionError: (3607, 1) != 3607
```


I am unsure how would I validate this output manually especially for analysis timeseries counts as expected count calculated by me just has total number of values from the input files.

**Comment by shankari** [comment]
> I am unsure how would I validate this output manually especially for analysis timeseries counts as expected count calculated by me just has total number of values from the input files.

You cannot validate it - is is probably incorrect. I would print out the key of the single analysis result. I bet some other test is not cleaning up properly after itself.

**Comment by shankari** [comment]
Ah the `modellingTests` are enabled! But alas, they are failing.

**Comment by shankari** [comment]
Let's get the tests to pass before I review. There's no point in reviewing PRs that don't pass tests.

**Comment by MukuFlash03** [comment]
> > I am unsure how would I validate this output manually especially for analysis timeseries counts as expected count calculated by me just has total number of values from the input files.
> 
> You cannot validate it - is is probably incorrect. I would print out the key of the single analysis result. I bet some other test is not cleaning up properly after itself, which is why you are running into this only when running in concert with other tests.

This is the entire analysis document entry being fetched on running all tests:
`key = segmentation/raw_stop`

`Entry({'_id': ObjectId('64fbb82b7b555615e043ef49'), 'user_id': UUID('3f205663-671b-3333-91d8-347d74a88f83'), 'metadata': {'key': 'segmentation/raw_stop', 'platform': 'server', 'write_ts': 1694218283.960188, 'time_zone': 'America/Los_Angeles', 'write_local_dt': {'year': 2023, 'month': 9, 'day': 8, 'hour': 17, 'minute': 11, 'second': 23, 'weekday': 4, 'timezone': 'America/Los_Angeles'}, 'write_fmt_time': '2023-09-08T17:11:23.960188-07:00'}, 'data': {'enter_ts': 5, 'exit_ts': 6, 'trip_id': 'test_trip_id'}})`


The specific test case I was testing was passing an empty key list to aggregate subclass.
If this invalid data isn't related to my implementation, for now can I ignore this test and commit code changes without this specific test?
As it does seem to work with empty list, almost matching the value and aggregating entries except for this extra entry.

**Comment by shankari** [comment]
I appreciate your desire to get this checked in soon, but the project is at a stage of maturity where I don't want to cut corners. And commenting out a test that doesn't work the way you want without understanding the reason why is optimizing the specific functionality you are working on over the health of the project as a whole.

>  If this invalid data isn't related to my implementation, for now can I ignore this test case and commit code changes without this specific test?

How do you know that this is not related to your implementation? You don't even know where the entry is coming from. Maybe it is actually correct. You have validated the `background/location` and `background/filtered_location` counts by using grep on the files. But have you tried to grep for any other keys in the files?

1. We are not under time pressure for these changes.
2. There is no external deadline we must meet.
3. We are in the process of improving the tightness and production-readiness of the codebase
4. There is no other magic team member who is going to go in and understand this new entry

We should understand where the entry is coming from and fix it, not fix by removing an inconvenient test

File: emission/tests/storageTests/TestTimeSeries.py [hunk 10]
```diff
@@ -81,6 +83,114 @@ def testExtraQueries(self):
...
+            self.assertEqual(count_ts8, 3607)
+        except AssertionError as e:
+            print(f"Assertion failed for 3607...")
+            for ct in count_ts8:
+                cte = ecwe.Entry(ct)
+                print(f"CTE = ")
+                print(cte.user_id)
+                print(cte.metadata.key)
+                print(cte)
+                print("=== Trip:", cte.data.start_loc, "->", cte.data.end_loc)
```
Comment (shankari) [inline emission/tests/storageTests/TestTimeSeries.py:181]: This will mean that this test can never fail, so it is essentially the same as the test being commented out like it was before. It is true that you print out an error message if you catch the assertion, but who is checking for log statements in the automated tests, particularly while running in a Ci/CD environment? 

The only thing that people pay attention to is the red or green badge 😄 

If you have actually fixed the test, you should have confidence in it and fail the test if the assertion fails.
You can do this by removing the try/catch or by leaving this in for debugging, but rethrowing the error after printing out the CTE

If you do choose to keep the try/catch + rethrow, you should add a TODO: to remove it in Sept 2024, so that we don't have any more vestigial code sitting around.

**Review by shankari — APPROVED** [review]
I am merging this for now because:
1. you have investigated and fixed the issue, and
2. there is a release train leaving tonight

However, I would like you to include the cleanup below in your next PR.
Please keep the comment unresolved until I check the next PR and resolve it
